### PR TITLE
DAOS-14535 pool: start/stop ds_pool_child individually

### PR DESCRIPTION
**Pull request by NiuYawei**
Reorganize ds_pool_child start/stop code to make it able to be
started/stopped individually:

- Introduced four states for ds_pool_child: NEW, STARTING, STARTED
  and STOPPING.
- ds_pool_child is added/removed to/from cache through the collective
  call of pool_child_add/delete_one(), now the cache doesn't hold the
  ds_pool_child reference anymore.
- Introduced ds_pool_child_start/stop() for callers to start/stop
  ds_pool_child individually, ds_pool_child_state() to query the state
  of ds_pool_child.
- Adjusted ds_pool_child_lookup() & ds_pool_child_get() a bit;

Required-githooks: true

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/pool/srv_target.c [hunk 1]
```diff
@@ -80,16 +80,29 @@ ds_pool_child_lookup(const uuid_t uuid)
...
-			child->spc_ref++;
 			return child;
 		}
 	}
 	return NULL;
 }
 
 struct ds_pool_child *
-ds_pool_child_get(struct ds_pool_child *child)
+ds_pool_child_get(const uuid_t uuid)
```
Comment (liw) [inline src/pool/srv_target.c:90]: How about adding a static `pool_child_lookup_noref` (or `ds_pool_child_lookup_noref` if you need an external one) and keeping the semantics of `ds_pool_child_lookup` unchanged (except for checking the state)? And we could remove `ds_pool_child_get` if necessary.

That seems to have less impact on other modules, and also more consistent with other "lookup/get/put" APIs we have.

File: src/pool/srv_util.c [hunk 2]
```diff
@@ -973,7 +973,6 @@ check_pool_targets(uuid_t pool_id, int *tgt_ids, int tgt_cnt, bool reint,
 	}
 
 	ABT_rwlock_unlock(pool->sp_lock);
-	ds_pool_child_put(pool_child);
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/pool/srv_util.c:976]: indeed, it's better to call ds_pool_child_get() to hold reference. Can I fix it in the follow-on PR (along with the two spelling issues.)?

File: src/pool/srv_target.c (hunk 1 above)
Comment (NiuYawei) [reply src/pool/srv_target.c:90]: Ok, sounds good to me. I'll fix it.

File: src/pool/srv_util.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (liw) [reply src/pool/srv_util.c:976]: Fine with me.

**Review by wangshilong — APPROVED** [review]

**Review by liw — APPROVED** [review]

File: src/pool/srv_target.c [hunk 3]
```diff
@@ -99,23 +112,9 @@ ds_pool_child_put(struct ds_pool_child *child)
...
-		ds_stop_chkpt_ult(child);
-		/* only stop gc ULT when all ops ULTs are done */
-		stop_gc_ult(child);
-		stop_flush_ult(child);
-
-		vos_pool_close(child->spc_hdl);
-		dss_module_fini_metrics(DAOS_TGT_TAG, child->spc_metrics);
-		ABT_eventual_set(child->spc_ref_eventual,
-				 (void *)&child->spc_ref,
+	if (child->spc_ref == 0 && *child->spc_state == POOL_CHILD_STOPPING)
```
Comment (liuxuezhao) [inline src/pool/srv_target.c:115]: seems when the spc_ref drop to zero, the spc_state must be STOPPING?

**Review by liuxuezhao — APPROVED** [review]

File: src/pool/srv_target.c (hunk 3 above)
Comment (NiuYawei) [reply src/pool/srv_target.c:115]: pool child cache doesn't hold pool child reference now, so the pool child reference could be dropped to zero even if it's still in cache (in start mode).

**Pull request merged**